index.js

```diff
@@ -25,24 +52,49 @@ function fastifyMiddie (fastify, options, next) {
     } else {
       this[kMiddie].use(path, fn)
     }
+    this[kMiddieHasMiddlewares] = true
     return this
   }
 
-  function runMiddie (req, reply, next) {
-    if (this[kMiddlewares].length > 0) {
-      req.raw.originalUrl = req.raw.url
-      req.raw.id = req.id
-      req.raw.hostname = req.hostname
-      req.raw.protocol = req.protocol
-      req.raw.ip = req.ip
-      req.raw.ips = req.ips
-      req.raw.log = req.log
-      req.raw.body = req.body
-      req.raw.query = req.query
-      reply.raw.log = req.log
-      this[kMiddie].run(req.raw, reply.raw, next)
+  function runMiddieByHook (hook) {
+    if (
+      supportedHooksWithPayload.includes(hook)
+    ) {
+      return function runMiddie (req, reply, _payload, next) {
+        if (this[kMiddieHasMiddlewares]) {
+          req.raw.originalUrl = req.raw.url
+          req.raw.id = req.id
+          req.raw.hostname = req.hostname
+          req.raw.protocol = req.protocol
+          req.raw.ip = req.ip
+          req.raw.ips = req.ips
+          req.raw.log = req.log
+          req.raw.body = req.body
+          req.raw.query = req.query
+          reply.raw.log = req.log
+          this[kMiddie].run(req.raw, reply.raw, next)
+        } else {
+          next()
+        }
```

Comment:
this code is equal to the the one in the `else condition`

I would just move the `function runMiddie (req, reply, _payload, next) {` to the global scope and then:

```
return runMiddie
else{
  return function runMiddieNoPayload (req, reply, next) {
     runMiddie(req, reply, null, next)
  }
}
```